### PR TITLE
feat: Add `NoiseSensorDeviceDetails`

### PR DESCRIPTION
**Pull request by xplato**
Adds read-only `NoiseSensorDeviceDetails` component.

Screenshot 2024-05-01 at 12 01 31 PM

### CODE REVIEW

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                      want to allow.
+                    </span>
+                    <span className='seam-tooltip-content'>
+                      Quiet hours is a separate threshold that takes effect only
+                      for a specified time range.
+                    </span>
+                  </div>
+                ) : (
+                  'A noise threshold is the highest noise level (in dB) you want to allow for a given time range in the day.'
+                )
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx:67]: Move all of these to `t.*`

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+              }
+            >
+              {!isInitialLoading && (noiseThresholds?.length ?? 0) > 0 ? (
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx:70]: Let's move all of the noise thresholds to a separate component so the loading logic, etc. is isolated.

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    </div>
+  )
+}
+
+const generateTimeZoneCaption = (
+  device: NoiseSensorDevice,
+  thresholds: NoiseThresholds[] | undefined
+): string | null => {
+  if (device.location?.timezone != null) {
+    return `All times in ${formatTimeZone(device.location.timezone)}`
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx:133]: Move to `t.*`

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  if (thresholds == null) {
+    return null
+  }
+
+  const timeZone =
+    thresholds[0]?.starts_daily_at.match(/\[(.*?)\]/)?.[1] ?? 'UTC'
+  return `All times in ${formatTimeZone(timeZone)}`
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx:142]: Move to `t.*`.

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+  if (device.location?.timezone != null) {
+    return `All times in ${formatTimeZone(device.location.timezone)}`
+  }
+
+  if (thresholds == null) {
+    return null
+  }
+
+  const timeZone =
+    thresholds[0]?.starts_daily_at.match(/\[(.*?)\]/)?.[1] ?? 'UTC'
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx:141]: What is this match for?

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            disableConnectedAccountInformation
+          }
+          disableResourceIds={disableResourceIds}
+        />
+      </div>
+    </div>
+  )
+}
+
+const generateTimeZoneCaption = (
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx:128]: ``
```suggestion
const getTimeZoneCaption = (
```

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                <DeviceModel device={device} />
+              </div>
+            </div>
+          </div>
+        </div>
+
+        <DetailSectionGroup>
+          <div className='seam-detail-section-wrap'>
+            <DetailSection
+              label='Noise thresholds'
```
Comment (razor-x) [inline src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx:52]: Move to `t.*`.

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx (hunk 2 above)
Comment (xplato) [reply src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx]: Due to the architecture of the UI, I had to move quite a lot, but it is now isolated!

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx (hunk 5 above)
Comment (xplato) [reply src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx]: Ahh, I will add a comment. It matches the timezone in the square brackets of the `starts_daily_at` string. Example: `'06:00:00[America/Los_Angeles]'`

Response value can be seen in the docs here: https://docs.seam.co/latest/api-clients/noise-sensors/list-noise-thresholds#list-noise-thresholds-set-on-device

Comment (razor-x) [reply src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx]: Oh yea, these are `seam-tod`. Can you try to parse it with this lib? I believe the API response should always be standardized to this: https://github.com/seamapi/zoned-time

Comment (razor-x) [reply src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx]: zoned-time is the successor to seam-tod, but the strings returned by the API should be compatible with zoned-time.

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+                </DetailRow>
+              ))
+            ) : (
+              <DetailRow
+                label={
+                  <span className='seam-detail-row-empty-label'>{t.none}</span>
+                }
+              />
+            )}
+          </DetailSection>
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:77]: This ternay structure is hard to parse. I was hoping that by moving it to a component, we can have a more linear, easy to follow style, e.g.


```
  const { noiseThresholds, isInitialLoading } = useNoiseThresholds({
    device_id: device.device_id,
  })
if (isInitialLoading) return null
if (noiseThresholds == null) return null
return noiseThresholds.map(noiseThreshold => <NoiseThreshold key={noiseThreshold.noise_threshold_id} noiseThreshold={noiseThreshold} />

File: src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx (hunk 5 above)
Comment (xplato) [reply src/lib/seam/components/DeviceDetails/NoiseSensorDeviceDetails.tsx]: Ahh, nice! Done.

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+export function NoiseThresholdsList({
+  device,
+}: NoiseThresholdsListProps): JSX.Element {
+  const { noiseThresholds, isInitialLoading } = useNoiseThresholds({
+    device_id: device.device_id,
+  })
+
+  return (
+    <>
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:23]:
```suggestion
```

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            <div className='seam-empty-div' />
+            <div className='seam-detail-section-footer-content'>
+              <div className='seam-detail-section-footer-content-text'>
+                <p>{getTimeZoneCaption(device, noiseThresholds)}</p>
+              </div>
+            </div>
+          </div>
+        </div>
+      </DetailSectionGroup>
+    </>
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:59]:
```suggestion
```

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  })
+
+  return (
+    <>
+      <DetailSectionGroup>
+        <div className='seam-detail-section-wrap'>
+          <DetailSection
+            label={t.noiseThresholds}
+            tooltipContent={
+              device.device_type === 'minut_sensor' ? (
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:29]: Why does `minut` have special behavior here?

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+import { DetailSectionGroup } from 'lib/ui/layout/DetailSectionGroup.js'
+
+interface NoiseThresholdsListProps {
+  device: NoiseSensorDevice
+}
+
+export function NoiseThresholdsList({
+  device,
+}: NoiseThresholdsListProps): JSX.Element {
+  const { noiseThresholds, isInitialLoading } = useNoiseThresholds({
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:18]: Why are we passing `isInitialLoading` around instead of showing a loading state for the entire section?

Are we using or can we use the loading behavior / loading toasts like we do elsewhere? It's ok if we address this as a follow up PR, but I thought we had standardized the component loading behavior.

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    </DetailRow>
+  ))
+}
+
+const getTimeZoneCaption = (
+  device: NoiseSensorDevice,
+  thresholds: NoiseThresholds[] | undefined
+): string | null => {
+  if (device.location?.timezone != null) {
+    return `All times in ${formatTimeZone(device.location.timezone)}`
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:118]:
```suggestion
    return `${t.allTimesIn} {formatTimeZone(device.location.timezone)}`
```

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+  }
+
+  if (thresholds == null) {
+    return null
+  }
+
+  const zonedTime = ZonedTime.from(thresholds[0]?.starts_daily_at ?? '')
+  return `All times in ${formatTimeZone(zonedTime.timeZone)}`
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:126]:
```suggestion
  return `${t.allTimesIn} ${formatTimeZone(zonedTime.timeZone)}`
```

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+): string | null => {
+  if (device.location?.timezone != null) {
+    return `All times in ${formatTimeZone(device.location.timezone)}`
+  }
+
+  if (thresholds == null) {
+    return null
+  }
+
+  const zonedTime = ZonedTime.from(thresholds[0]?.starts_daily_at ?? '')
```
Comment (razor-x) [inline src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:125]: I think ZonedTime will throw an error if you pass null or an empty string.
https://github.com/seamapi/zoned-time/blob/main/src/lib/parse-zoned-time-string.ts#L24-L32

This should be a type error, but if not it means there is an upstream type issue in zoned time. Feel free to PR upstream if you find it.

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx (hunk 11 above)
Comment (xplato) [reply src/lib/ui/noise-sensor/NoiseThresholdsList.tsx]: Not sure—it is specified in the design: https://www.figma.com/file/WTdzEh4CudGdQFOtuVR3Eh/Thermostat%2FNoise-components?type=design&node-id=3996-10671&mode=design&t=GfLxqO8DBwAaTIaZ-4

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx (hunk 12 above)
Comment (xplato) [reply src/lib/ui/noise-sensor/NoiseThresholdsList.tsx:18]: I wanted to avoid an additional ternary operator on `Content`. But that's a good point, I can add the loading toast as well. There's no design for this, and it will overlay the device card at the top. I'll address this in a subsequent PR if you don't mind.

File: src/lib/ui/noise-sensor/NoiseThresholdsList.tsx (hunk 15 above)
Comment (xplato) [reply src/lib/ui/noise-sensor/NoiseThresholdsList.tsx]: Nope, all good—added an additional check on the first threshold to resolve this.

File: .storybook/seed-fake.js [hunk 16]
```diff
@@ -468,6 +468,42 @@ export const seedFake = (db) => {
...
+
+  db.addNoiseThreshold({
+    device_id: device7.device_id,
+    workspace_id: ws2.workspace_id,
+    created_at: '2023-05-17T00:16:12.000',
+    noise_threshold_decibels: 75,
+    starts_daily_at: '06:00:00[America/Los_Angeles]',
+    ends_daily_at: '22:00:00[America/Los_Angeles]',
+    name: 'builtin_quiet_hours',
+  })
```
Comment (razor-x) [inline .storybook/seed-fake.js:505]: The underscore names look a little odd in the storybook. Can we do this? 

```suggestion
  db.addNoiseThreshold({
    device_id: device7.device_id,
    workspace_id: ws2.workspace_id,
    created_at: '2023-05-17T00:16:12.000',
    noise_threshold_decibels: 70,
    starts_daily_at: '22:00:00[America/Los_Angeles]',
    ends_daily_at: '06:00:00[America/Los_Angeles]',
    name: 'Quiet Hours',
  })

  db.addNoiseThreshold({
    device_id: device7.device_id,
    workspace_id: ws2.workspace_id,
    created_at: '2023-05-17T00:16:12.000',
    noise_threshold_decibels: 75,
    starts_daily_at: '06:00:00[America/Los_Angeles]',
    ends_daily_at: '22:00:00[America/Los_Angeles]',
    name: 'Active Hours',
  })
```

**Review by razor-x — APPROVED** [review]
